File: src/primitives/workspace.ts
/**
 * Workspace Primitive
 *
 * The top-level container for agent collaboration. A workspace contains
 * channels, manages agents, and coordinates the overall communication environment.
 */

import { nanoid } from 'nanoid';
import { EventEmitter } from 'eventemitter3';
import { Channel, ChannelType, createDMChannel, createGroupDM } from './channel.js';
import { Message, MessageType } from './message.js';
import { Thread, ThreadManager } from './thread.js';
import { ReactionManager } from './reaction.js';
import { PresenceManager, PresenceStatus } from './presence.js';
import { ArtifactStore } from './artifact.js';
import { parseMentions, MentionContext } from './mention.js';

/**
 * Configuration for a workspace.
 */
export interface WorkspaceSettings {
  maxMessageLength: number;
  allowMessageEditing: boolean;
  messageRetentionDays?: number;
  defaultChannelType: ChannelType;
  allowPublicChannels: boolean;
  requireAgentApproval: boolean;
  maxAgents?: number;
  defaultNotifications: 'all' | 'mentions' | 'none';
}

const DEFAULT_SETTINGS: WorkspaceSettings = {
  maxMessageLength: 40000,
  allowMessageEditing: true,
  defaultChannelType: ChannelType.PUBLIC,
  allowPublicChannels: true,
  requireAgentApproval: false,
  defaultNotifications: 'mentions',
};

/**
 * Agent identity within a workspace.
 */
export interface WorkspaceAgent {
  id: string;
  name: string;
  description: string;
  avatarEmoji: string;
  capabilities: string[];
  joinedAt: Date;
  channelIds: string[];
}

/**
 * Events emitted by the workspace.
 */
export interface WorkspaceEvents {
  message: (message: Message, context: MentionContext) => void;
  reaction: (messageId: string, emoji: string, agentId: string) => void;
  agentJoined: (agentId: string) => void;
  agentLeft: (agentId: string) => void;
  channelCreated: (channel: Channel) => void;
  channelJoined: (channelId: string, agentId: string) => void;
  channelLeft: (channelId: string, agentId: string) => void;
  threadCreated: (thread: Thread) => void;
  presenceChanged: (agentId: string, status: PresenceStatus) => void;
}

/**
 * A workspace for agent collaboration.
 */
export class AgentWorkspace extends EventEmitter<WorkspaceEvents> {
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly iconEmoji: string;
  readonly settings: WorkspaceSettings;
  readonly createdAt: Date;

  // Managers
  readonly threads: ThreadManager;
  readonly reactions: ReactionManager;
  readonly presence: PresenceManager;
  readonly artifacts: ArtifactStore;

  // Registries
  private readonly _agents: Map<string, WorkspaceAgent> = new Map();
  private readonly _agentsByName: Map<string, string> = new Map();
  private readonly _channels: Map<string, Channel> = new Map();
  private readonly _channelsByName: Map<string, string> = new Map();

  constructor(
    name: string,
    options: {
      description?: string;
      iconEmoji?: string;
      settings?: Partial<WorkspaceSettings>;
    } = {}
  ) {
    super();

    this.id = nanoid();
    this.name = name;
    this.description = options.description ?? '';
    this.iconEmoji = options.iconEmoji ?? '🏢';
    this.settings = { ...DEFAULT_SETTINGS, ...options.settings };
    this.createdAt = new Date();

    // Initialize managers
    this.threads = new ThreadManager();
    this.reactions = new ReactionManager();
    this.presence = new PresenceManager();
    this.artifacts = new ArtifactStore();

    // Create default #general channel
    this.createChannel('general', { purpose: 'General discussion' });
  }

  // ==================== Agent Management ====================

  /**
   * Register an agent to the workspace.
   */
  registerAgent(agent: {
    id: string;
    name: string;
    description?: string;
    avatarEmoji?: string;
    capabilities?: string[];
  }): WorkspaceAgent {
    if (this.settings.maxAgents && this._agents.size >= this.settings.maxAgents) {
      throw new Error(`Workspace has reached max agents (${this.settings.maxAgents})`);
    }

    const workspaceAgent: WorkspaceAgent = {
      id: agent.id,
      name: agent.name,
      description: agent.description ?? '',
      avatarEmoji: agent.avatarEmoji ?? '🤖',
      capabilities: agent.capabilities ?? [],
      joinedAt: new Date(),
      channelIds: [],
    };

    this._agents.set(agent.id, workspaceAgent);
    this._agentsByName.set(agent.name.toLowerCase(), agent.id);

    // Initialize presence
    this.presence.getOrCreate(agent.id);

    // Auto-join #general
    const general = this.getChannel('general');
    if (general) {
      this.joinChannel(agent.id, general.id);
    }

    this.emit('agentJoined', agent.id);
    return workspaceAgent;
  }

  /**
   * Get an agent by ID.
   */
  getAgent(agentId: string): WorkspaceAgent | undefined {
    return this._agents.get(agentId);
  }

  /**
   * Get an agent by name (for @mention lookup).
   */
  getAgentByName(name: string): WorkspaceAgent | undefined {
    const agentId = this._agentsByName.get(name.toLowerCase());
    return agentId ? this._agents.get(agentId) : undefined;
  }

  /**
   * List all agents, optionally filtered by status.
   */
  listAgents(options: { status?: PresenceStatus } = {}): WorkspaceAgent[] {
    let agents = Array.from(this._agents.values());

    if (options.status) {
      const agentIds = new Set(this.presence.getByStatus(options.status));
      agents = agents.filter((a) => agentIds.has(a.id));
    }

    return agents;
  }

  /**
   * Find all agents with a specific capability.
   */
  findAgentsByCapability(capability: string): WorkspaceAgent[] {
    return Array.from(this._agents.values()).filter((a) =>
      a.capabilities.includes(capability)
    );
  }

  // ==================== Channel Management ====================

  /**
   * Create a new channel in the workspace.
   */
  createChannel(
    name: string,
    options: {
      purpose?: string;
      channelType?: ChannelType;
      creatorId?: string;
    } = {}
  ): Channel {
    const normalizedName = name.toLowerCase().replace(/\s+/g, '-');

    if (this._channelsByName.has(normalizedName)) {
      throw new Error(`Channel #${normalizedName} already exists`);
    }

    const channel = new Channel({
      name: normalizedName,
      purpose: options.purpose,
      channelType: options.channelType ?? this.settings.defaultChannelType,
      workspaceId: this.id,
    });

    if (options.creatorId) {
      channel.addMember(options.creatorId, 'admin');
    }

    this._channels.set(channel.id, channel);
    this._channelsByName.set(normalizedName, channel.id);

    this.emit('channelCreated', channel);
    return channel;
  }

  /**
   * Get a channel by name or ID.
   */
  getChannel(nameOrId: string): Channel | undefined {
    // Try by ID first
    if (this._channels.has(nameOrId)) {
      return this._channels.get(nameOrId);
    }

    // Try by name
    const channelId = this._channelsByName.get(nameOrId.replace(/^#/, '').toLowerCase());
    return channelId ? this._channels.get(channelId) : undefined;
  }

  /**
   * List channels in the workspace.
   */
  listChannels(options: { includePrivate?: boolean; agentId?: string } = {}): Channel[] {
    const { includePrivate = false, agentId } = options;
    const channels: Channel[] = [];

    for (const channel of this._channels.values()) {
      if (channel.isDM) continue;

      if (channel.channelType === ChannelType.PRIVATE) {
        if (!includePrivate) continue;
        if (agentId && !channel.isMember(agentId)) continue;
      }

      channels.push(channel);
    }

    return channels.sort((a, b) => a.name.localeCompare(b.name));
  }

  /**
   * Have an agent join a channel.
   */
  joinChannel(agentId: string, channelId: string): boolean {
    const agent = this._agents.get(agentId);
    const channel = this._channels.get(channelId);

    if (!agent || !channel) return false;

    channel.addMember(agentId);
    agent.channelIds.push(channelId);

    this.emit('channelJoined', channelId, agentId);
    return true;
  }

  /**
   * Have an agent leave a channel.
   */
  leaveChannel(agentId: string, channelId: string): boolean {
    const agent = this._agents.get(agentId);
    const channel = this._channels.get(channelId);

    if (!agent || !channel) return false;

    channel.removeMember(agentId);
    agent.channelIds = agent.channelIds.filter((id) => id !== channelId);

    this.emit('channelLeft', channelId, agentId);
    return true;
  }

  /**
   * Get or create a DM channel between two agents.
   */
  getDMChannel(agent1Id: string, agent2Id: string): Channel {
    // Check if DM already exists
    for (const channel of this._channels.values()) {
      if (channel.channelType === ChannelType.DIRECT) {
        const members = new Set(channel.memberIds);
        if (members.has(agent1Id) && members.has(agent2Id) && members.size === 2) {
          return channel;
        }
      }
    }

    // Create new DM
    const channel = createDMChannel(agent1Id, agent2Id);
    this._channels.set(channel.id, channel);
    return channel;
  }

  /**
   * Create a group DM.
   */
  createGroupDM(agentIds: string[], name?: string): Channel {
    const channel = createGroupDM(agentIds, name);
    this._channels.set(channel.id, channel);
    return channel;
  }

  // ==================== Messaging ====================

  /**
   * Post a message to a channel.
   */
  postMessage(
    senderId: string,
    channelId: string,
    content: string,
    options: {
      messageType?: MessageType;
      threadId?: string;
      artifactIds?: string[];
      metadata?: Record<string, unknown>;
    } = {}
  ): Message {
    const agent = this._agents.get(senderId);
    const channel = this._channels.get(channelId);

    if (!agent) {
      throw new Error(`Agent ${senderId} not found`);
    }
    if (!channel) {
      throw new Error(`Channel ${channelId} not found`);
    }
    if (!channel.isMember(senderId)) {
      throw new Error(`Agent ${senderId} is not a member of channel`);
    }
    if (content.length > this.settings.maxMessageLength) {
      throw new Error(`Message exceeds max length (${this.settings.maxMessageLength})`);
    }

    const message = new Message({
      senderId,
      channelId,
      content,
      messageType: options.messageType ?? MessageType.TEXT,
      threadId: options.threadId,
      artifactIds: options.artifactIds,
      metadata: options.metadata,
    });

    // Handle threading
    if (options.threadId) {
      const thread = this.threads.getThread(options.threadId);
      if (thread) {
        thread.addReply(message);
      }
    }

    // Post to channel
    channel.postMessage(message);

    // Update presence
    const presence = this.presence.get(senderId);
    presence?.recordActivity();

    // Parse mentions and create context
    const mentions = parseMentions(content, {
      resolveAgent: (name) => this._agentsByName.get(name.toLowerCase()),
      resolveChannel: (name) => this._channelsByName.get(name.toLowerCase()),
    });

    const mentionContext = new MentionContext({
      messageId: message.id,
      channelId,
      senderId,
      mentions,
    });

    this.emit('message', message, mentionContext);
    return message;
  }

  /**
   * Reply to a message in a thread.
   */
  replyInThread(
    senderId: string,
    parentMessageId: string,
    content: string,
    options: {
      artifactIds?: string[];
      metadata?: Record<string, unknown>;
    } = {}
  ): Message {
    // Find the parent message
    let parentMessage: Message | undefined;
    let channel: Channel | undefined;

    for (const ch of this._channels.values()) {
      const msg = ch.getMessage(parentMessageId);
      if (msg) {
        parentMessage = msg;
        channel = ch;
        break;
      }
    }

    if (!parentMessage || !channel) {
      throw new Error(`Parent message ${parentMessageId} not found`);
    }

    // Get or create thread
    const thread = this.threads.getOrCreateThread(parentMessage);

    // Create the reply
    const reply = new Message({
      senderId,
      channelId: channel.id,
      content,
      threadId: thread.id,
      artifactIds: options.artifactIds,
      metadata: options.metadata,
    });

    thread.addReply(reply);

    // Parse mentions
    const mentions = parseMentions(content, {
      resolveAgent: (name) => this._agentsByName.get(name.toLowerCase()),
      resolveChannel: (name) => this._channelsByName.get(name.toLowerCase()),
    });

    const mentionContext = new MentionContext({
      messageId: reply.id,
      channelId: channel.id,
      senderId,
      mentions,
    });

    this.emit('message', reply, mentionContext);
    return reply;
  }

  /**
   * Add a reaction to a message.
   */
  addReaction(messageId: string, emoji: string, agentId: string) {
    const reaction = this.reactions.addReaction(messageId, emoji, agentId);
    this.emit('reaction', messageId, emoji, agentId);
    return reaction;
  }

  /**
   * Get messages from a channel.
   */
  getMessages(
    channelId: string,
    options: { limit?: number; before?: Date; after?: Date } = {}
  ): Message[] {
    const channel = this._channels.get(channelId);
    return channel?.getMessages(options) ?? [];
  }

  // ==================== Presence ====================

  /**
   * Update agent presence.
   */
  setPresence(agentId: string, status: PresenceStatus, message = '') {
    const presence = this.presence.update(agentId, status, message);
    this.emit('presenceChanged', agentId, status);
    return presence;
  }

  /**
   * Get online agents.
   */
  getOnlineAgents(): WorkspaceAgent[] {
    const onlineIds = new Set(this.presence.getOnlineAgents());
    return Array.from(this._agents.values()).filter((a) => onlineIds.has(a.id));
  }

  /**
   * Get available agents (online with capacity).
   */
  getAvailableAgents(): WorkspaceAgent[] {
    const availableIds = new Set(this.presence.getAvailableAgents());
    return Array.from(this._agents.values()).filter((a) => availableIds.has(a.id));
  }

  // ==================== Serialization ====================

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      iconEmoji: this.iconEmoji,
      agentCount: this._agents.size,
      channelCount: this._channels.size,
      createdAt: this.createdAt,
    };
  }

  toString(): string {
    return `${this.iconEmoji} ${this.name} (${this._agents.size} agents, ${this._channels.size} channels)`;
  }
}

File: agent_slack/primitives/workspace.py
"""
Workspace Primitive

The top-level container for agent collaboration. A workspace contains
channels, agents, and manages the overall communication environment.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Callable, Any
import uuid

from .agent import Agent, AgentStatus
from .channel import Channel, ChannelType, create_dm_channel, create_group_dm
from .message import Message, MessageType


@dataclass
class WorkspaceSettings:
    """Configuration for a workspace."""
    
    # Message settings
    max_message_length: int = 40000
    allow_message_editing: bool = True
    message_retention_days: Optional[int] = None  # None = forever
    
    # Channel settings
    default_channel_type: ChannelType = ChannelType.PUBLIC
    allow_public_channels: bool = True
    
    # Agent settings
    require_agent_approval: bool = False
    max_agents: Optional[int] = None
    
    # Notification settings
    default_notifications: str = "mentions"  # "all", "mentions", "none"


@dataclass
class Workspace:
    """
    A workspace for agent collaboration.
    
    Key concepts:
    - Workspaces are isolated environments (like Slack workspaces)
    - Agents must be registered to participate
    - Channels live within workspaces
    - All messages flow through the workspace
    
    Design decisions:
    - Workspace is the "god object" that manages everything
    - In production, this would be backed by a database
    - Event hooks allow extensibility (logging, webhooks, etc.)
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Identity
    name: str = ""
    description: str = ""
    icon_emoji: str = "🏢"
    
    # Settings
    settings: WorkspaceSettings = field(default_factory=WorkspaceSettings)
    
    # Registries
    agents: dict[str, Agent] = field(default_factory=dict)
    channels: dict[str, Channel] = field(default_factory=dict)
    channels_by_name: dict[str, str] = field(default_factory=dict)  # name -> id
    
    # Timestamps
    created_at: datetime = field(default_factory=datetime.utcnow)
    
    # Event hooks
    _message_hooks: list[Callable[[Message], Any]] = field(default_factory=list, repr=False)
    _event_hooks: list[Callable[[str, dict], Any]] = field(default_factory=list, repr=False)
    
    def __post_init__(self):
        """Initialize workspace with default channel."""
        if not self.name:
            self.name = f"workspace-{self.id[:8]}"
        
        # Create default #general channel
        self.create_channel("general", purpose="General discussion")
    
    # ==================== Agent Management ====================
    
    def register_agent(self, agent: Agent) -> Agent:
        """
        Register an agent to the workspace.
        
        Agents must be registered before they can participate.
        """
        if self.settings.max_agents and len(self.agents) >= self.settings.max_agents:
            raise ValueError(f"Workspace has reached max agents ({self.settings.max_agents})")
        
        agent.workspace_id = self.id
        self.agents[agent.id] = agent
        
        # Auto-join default channels
        general = self.get_channel("general")
        if general:
            self.join_channel(agent.id, general.id)
        
        self._emit_event("agent_registered", {"agent_id": agent.id})
        return agent
    
    def register(self, *agents: Agent) -> list[Agent]:
        """Register multiple agents at once."""
        return [self.register_agent(agent) for agent in agents]
    
    def get_agent(self, agent_id: str) -> Optional[Agent]:
        """Get an agent by ID."""
        return self.agents.get(agent_id)
    
    def get_agent_by_name(self, name: str) -> Optional[Agent]:
        """Get an agent by name (for @mention lookup)."""
        for agent in self.agents.values():
            if agent.name == name:
                return agent
        return None
    
    def list_agents(self, status: Optional[AgentStatus] = None) -> list[Agent]:
        """List all agents, optionally filtered by status."""
        agents = list(self.agents.values())
        if status:
            agents = [a for a in agents if a.status == status]
        return agents
    
    def find_agents_by_capability(self, capability: str) -> list[Agent]:
        """Find all agents with a specific capability."""
        return [a for a in self.agents.values() if a.has_capability(capability)]
    
    # ==================== Channel Management ====================
    
    def create_channel(
        self,
        name: str,
        purpose: str = "",
        channel_type: ChannelType = ChannelType.PUBLIC,
        creator_id: Optional[str] = None,
    ) -> Channel:
        """Create a new channel in the workspace."""
        # Normalize name
        name = name.lower().replace(" ", "-")
        
        if name in self.channels_by_name:
            raise ValueError(f"Channel #{name} already exists")
        
        channel = Channel(
            name=name,
            purpose=purpose,
            channel_type=channel_type,
            workspace_id=self.id,
        )
        
        # Add creator as admin if specified
        if creator_id:
            channel.add_member(creator_id, role="admin")
        
        self.channels[channel.id] = channel
        self.channels_by_name[name] = channel.id
        
        self._emit_event("channel_created", {
            "channel_id": channel.id,
            "name": name,
            "creator_id": creator_id,
        })
        
        return channel
    
    def get_channel(self, name_or_id: str) -> Optional[Channel]:
        """Get a channel by name or ID."""
        # Try by ID first
        if name_or_id in self.channels:
            return self.channels[name_or_id]
        
        # Try by name
        channel_id = self.channels_by_name.get(name_or_id.lstrip("#"))
        if channel_id:
            return self.channels[channel_id]
        
        return None
    
    def list_channels(
        self,
        include_private: bool = False,
        agent_id: Optional[str] = None,
    ) -> list[Channel]:
        """
        List channels in the workspace.
        
        If agent_id is specified, only show channels the agent can see.
        """
        channels = []
        
        for channel in self.channels.values():
            # Skip DMs
            if channel.is_dm:
                continue
            
            # Check visibility
            if channel.channel_type == ChannelType.PRIVATE:
                if not include_private:
                    continue
                if agent_id and not channel.is_member(agent_id):
                    continue
            
            channels.append(channel)
        
        return sorted(channels, key=lambda c: c.name)
    
    def join_channel(self, agent_id: str, channel_id: str) -> bool:
        """Have an agent join a channel."""
        agent = self.agents.get(agent_id)
        channel = self.channels.get(channel_id)
        
        if not agent or not channel:
            return False
        
        channel.add_member(agent_id)
        agent.join_channel(channel_id)
        
        self._emit_event("channel_joined", {
            "agent_id": agent_id,
            "channel_id": channel_id,
        })
        
        return True
    
    def leave_channel(self, agent_id: str, channel_id: str) -> bool:
        """Have an agent leave a channel."""
        agent = self.agents.get(agent_id)
        channel = self.channels.get(channel_id)
        
        if not agent or not channel:
            return False
        
        channel.remove_member(agent_id)
        agent.leave_channel(channel_id)
        
        self._emit_event("channel_left", {
            "agent_id": agent_id,
            "channel_id": channel_id,
        })
        
        return True
    
    def get_dm_channel(self, agent1_id: str, agent2_id: str) -> Channel:
        """Get or create a DM channel between two agents."""
        # Check if DM already exists
        for channel in self.channels.values():
            if channel.channel_type == ChannelType.DIRECT:
                members = set(channel.members.keys())
                if members == {agent1_id, agent2_id}:
                    return channel
        
        # Create new DM
        channel = create_dm_channel(agent1_id, agent2_id)
        channel.workspace_id = self.id
        self.channels[channel.id] = channel
        
        return channel
    
    # ==================== Messaging ====================
    
    def post_message(
        self,
        sender_id: str,
        channel_id: str,
        content: str,
        message_type: MessageType = MessageType.TEXT,
        **kwargs,
    ) -> Message:
        """
        Post a message to a channel.
        
        This is the main entry point for sending messages.
        """
        agent = self.agents.get(sender_id)
        channel = self.channels.get(channel_id)
        
        if not agent:
            raise ValueError(f"Agent {sender_id} not found")
        if not channel:
            raise ValueError(f"Channel {channel_id} not found")
        if not channel.is_member(sender_id):
            raise ValueError(f"Agent {sender_id} is not a member of channel")
        
        # Validate message length
        if len(content) > self.settings.max_message_length:
            raise ValueError(f"Message exceeds max length ({self.settings.max_message_length})")
        
        message = Message(
            sender_id=sender_id,
            channel_id=channel_id,
            content=content,
            message_type=message_type,
            **kwargs,
        )
        
        channel.post_message(message)
        
        # Update agent's last seen
        agent.last_seen = datetime.utcnow()
        
        # Run hooks
        for hook in self._message_hooks:
            hook(message)
        
        self._emit_event("message_posted", {
            "message_id": message.id,
            "sender_id": sender_id,
            "channel_id": channel_id,
        })
        
        return message
    
    def get_messages(
        self,
        channel_id: str,
        limit: int = 100,
        before: Optional[datetime] = None,
        after: Optional[datetime] = None,
    ) -> list[Message]:
        """Get messages from a channel."""
        channel = self.channels.get(channel_id)
        if not channel:
            return []
        
        return channel.get_messages(limit=limit, before=before, after=after)
    
    # ==================== Event System ====================
    
    def on_message(self, callback: Callable[[Message], Any]) -> None:
        """Register a callback for new messages."""
        self._message_hooks.append(callback)
    
    def on_event(self, callback: Callable[[str, dict], Any]) -> None:
        """Register a callback for workspace events."""
        self._event_hooks.append(callback)
    
    def _emit_event(self, event_type: str, data: dict) -> None:
        """Emit an event to all registered hooks."""
        for hook in self._event_hooks:
            hook(event_type, data)
    
    # ==================== Serialization ====================
    
    def to_dict(self) -> dict:
        """Serialize workspace to dictionary."""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "icon_emoji": self.icon_emoji,
            "agent_count": len(self.agents),
            "channel_count": len(self.channels),
            "created_at": self.created_at.isoformat(),
        }
    
    def __str__(self) -> str:
        """Human-readable representation."""
        return f"{self.icon_emoji} {self.name} ({len(self.agents)} agents, {len(self.channels)} channels)"
